src/ObjectUtils/PdbFileDia.cpp

```diff
@@ -0,0 +1,137 @@
+  }
+
+  return dia_data_source;
+}
+
+PdbFileDia::PdbFileDia(std::filesystem::path file_path, const ObjectFileInfo& object_file_info)
+    : file_path_(std::move(file_path)), object_file_info_(object_file_info) {}
+
+ErrorMessageOr<void> PdbFileDia::LoadDataForPDB() {
+  OUTCOME_TRY(dia_data_source_, CreateDiaDataSource());
+
+  // Open and prepare a program database (.pdb) file as a debug data source.
+  HRESULT result = dia_data_source_->loadDataFromPdb(file_path_.wstring().c_str());
+  if (result != S_OK) {
+    return ErrorMessage{
+        absl::StrFormat("loadDataFromPdb failed for %s (%u)", file_path_.string(), result)};
+  }
+
+  // Open DIA session.
+  result = dia_data_source_->openSession(&dia_session_);
+  if (result != S_OK) {
+    return ErrorMessage{
+        absl::StrFormat("openSession failed for %s (%u)", file_path_.string(), result)};
+  }
+
+  // Retrieve a reference to the global scope.
+  result = dia_session_->get_globalScope(&dia_global_scope_symbol_);
+  if (result != S_OK) {
+    return ErrorMessage{
+        absl::StrFormat("get_globalScope failed for %s (%u)", file_path_.string(), result)};
+  }
+
+  // Get age.
+  DWORD age = 0;
+  result = dia_global_scope_symbol_->get_age(&age);
+  if (result != S_OK) {
+    return ErrorMessage{absl::StrFormat("get_age failed for %s (%u)", file_path_.string(), result)};
+  }
+  age_ = age;
+
+  // Get guid.
+  GUID guid = {};
+  result = dia_global_scope_symbol_->get_guid(&guid);
+  if (result != S_OK) {
+    return ErrorMessage{
+        absl::StrFormat("get_guid failed for %s (%u)", file_path_.string(), result)};
+  }
+  CHECK(sizeof(guid_) == sizeof(GUID));
+  std::memcpy(guid_.data(), &guid, sizeof(GUID));
+
+  return outcome::success();
+}
+
+ErrorMessageOr<orbit_grpc_protos::ModuleSymbols> PdbFileDia::LoadDebugSymbols() {
+  HRESULT result = 0;
+  CComPtr<IDiaEnumSymbols> dia_enum_symbols;
+  result = dia_global_scope_symbol_->findChildren(SymTagFunction, NULL, nsNone, &dia_enum_symbols);
+  if (result != S_OK) {
+    return ErrorMessage{
+        absl::StrFormat("findChildren failed for %s (%u)", file_path_.string(), result)};
+  }
+
+  ModuleSymbols module_symbols;
+  module_symbols.set_load_bias(object_file_info_.load_bias);
+  module_symbols.set_symbols_file_path(file_path_.string());
+  CComPtr<IDiaSymbol> dia_symbol;
+  ULONG celt = 0;
+
+  while (SUCCEEDED(dia_enum_symbols->Next(1, &dia_symbol, &celt)) && (celt == 1)) {
+    SymbolInfo* symbol_info = module_symbols.add_symbol_infos();
+
+    BSTR function_name = {};
+    if (dia_symbol->get_name(&function_name) == S_OK) {
+      std::wstring name(function_name);
+      symbol_info->set_name(std::string(name.begin(), name.end()));
+      symbol_info->set_demangled_name(llvm::demangle(symbol_info->name()));
+      SysFreeString(function_name);
+    }
+
+    DWORD relative_virtual_address = 0;
+    if (dia_symbol->get_relativeVirtualAddress(&relative_virtual_address) == S_OK) {
```

Comment:
Shouldn't we skip entries entirely if the data is not complete? Otherwise, we end up with incorrect symbol entries in the UI, such as function names with address and/or size 0.

Reply:
Yeah good point, done, thanks.